runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPortablePipelineTranslator.java

```diff
@@ -301,6 +305,24 @@ private <K, V> void translateReshuffle(
         Iterables.getOnlyElement(transform.getOutputsMap().values()), inputDataStream.rebalance());
   }
 
+  private <K, V> void translateRedistributeByKey(
+      String id, RunnerApi.Pipeline pipeline, StreamingTranslationContext context) {
+    RunnerApi.PTransform transform = pipeline.getComponents().getTransformsOrThrow(id);
+    DataStream<WindowedValue<KV<K, V>>> inputDataStream =
+        context.getDataStreamOrThrow(Iterables.getOnlyElement(transform.getInputsMap().values()));
+    context.addDataStream(
+        Iterables.getOnlyElement(transform.getOutputsMap().values()), inputDataStream.rebalance());
```

Comment:
Where is the key being used here? (This implementation looks identical to the one below.)

Reply:
Looks identical to the code above as well. I suppose my preference is to not duplicate code unless there's a god reason for divergence or it's difficult to do so. E.g. above why not just write

```
translatorMap.put(PTransformTranslation.REDISTRIBUTE_BY_KEY_URN, this::translateReshuffle);
translatorMap.put(PTransformTranslation.REDISTRIBUTE_ARBITRARILY_URN, this::translateReshuffle);
```

which makes it clear that these are all implemented the same rather than trying to pattern match and see what the differences, if any, are.

It's not logical necessity, but saying "this (same) implementation is good enough for this usecase as well."

Reply:
But if you feel strongly, no need for this to be a blocker.

Reply:
Made a long post. I very much view their current "sameness" as coincidence that can diverge. I want to freeze Reshuffle where it is and Redistribute be the place that evolves to be different. (for example incorporating more of Jan's ideas for Flink, and using new Dataflow features for Dataflow)